include/openthread/tcp.h

```diff
@@ -0,0 +1,521 @@
+
+/**
+ * Binds the active socket to an IP address and port.
+ *
+ * @param[in]  aInstance   A pointer to the OpenThread instance to which the socket belongs.
+ * @param[in]  aSocket     A pointer to the active TCP socket structure to bind.
+ * @param[in]  aSockName   The address and port to which to bind this socket.
+ *
+ * @retval OT_ERROR_NONE    Successfully bound the active TCP socket.
+ * @retval OT_ERROR_FAILED  Failed to bind the active TCP socket.
+ */
+otError otTcpActiveBind(otInstance *aInstance, otTcpActiveSocket *aSocket, const otSockAddr *aSockName);
+
+/**
+ * This enumeration defines flags passed to @p otTcpActiveConnect.
+ */
+enum {
+    OT_TCP_ACTIVE_CONNECT_FAST_OPEN =    1 << 0,
+    OT_TCP_ACTIVE_CONNECT_NO_FAST_OPEN = 1 << 1,
+};
+
+/**
+ * Records the remote host and port for this connection. If TCP Fast Open is
+ * used (it can be enabled or disabled using @p aFlags), then the TCP
+ * connection establishment handshake is initiated immediately; otherwise, it
+ * is initiated when data is first added to the send buffer.
+ *
+ * @param[in]  aInstance  A pointer to the OpenThread instance to which the socket belongs.
+ * @param[in]  aSocket    A pointer to the active TCP socket structure to connect.
+ * @param[in]  aSockName  The IP address and port of the host to which to connect.
+ * @param[in]  aFlags     Flags specifying options for this operation (see enumeration above).
+ *
+ * @retval OT_ERROR_NONE    Successfully completed the operation.
+ * @retval OT_ERROR_FAILED  Failed to complete the operation.
+ */
+otError otTcpActiveConnect(otInstance *aInstance, otTcpActiveSocket *aSocket, const otSockAddr *aSockName, uint32_t aFlags);
+
+/**
+ * This enumeration defines flags passed to @p otTcpActiveSendByReference..
+ */
+enum {
+    OT_TCP_ACTIVE_SEND_MORE_TO_COME = 1 << 0,
+};
+
+/**
+ * Adds data referenced by the linked buffer chain pointed to by @p aBuffer to
+ * the send buffer. Upon a sucessful call to this function, the linked buffers
+ * and data they reference are owned by the TCP stack; they should not be
+ * modified by the application until a "send done" callback returns ownership
+ * of those objects to the application.
+ *
+ * @param[in]  aInstance  A pointer to the OpenThread instance to which the socket belongs.
+ * @param[in]  aSocket    A pointer to the active TCP socket structure representing the TCP endpoint on which to send data.
+ * @param[in]  aBuffer    A pointer to the linked buffer chain referencing data to add to the send buffer.
+ * @param[in]  aFlags     Flags specifying options for this operation (see enumeration above).
+ *
+ * @retval OT_ERROR_NONE    Successfully added data to the send buffer.
+ * @retval OT_ERROR_FAILED  Failed to add data to the send buffer.
+ */
+otError otTcpActiveSendByReference(otInstance *aInstance, otTcpActiveSocket *aSocket, otLinkedBuffer *aBuffer, uint32_t aFlags);
```

Comment:
I think we should disallow appending data once the `otLinkedBuffer` is in transmission (before "send done"), and one "send done" event should be sufficient for multiple `SendByReference`. 

Also, I think we can only append data if the `otLinkedBuffer` is the last `otLinkedBuffer` in send window.

Reply:
@abtink We *do* allow the user to call this function after a previous successful call to this function but before the send done callback. My intention was that each "send done" callback corresponds to a single `otLinkedBuffer` without any chaining. We could disallow chaining in `otTcpActiveSendByReference` if we want it to correspond one-to-one with calls to `otTcpActiveSendByReference`. This would still be compatible with Stuart's suggestion at our meeting today because we can support a "more to come" flag to prevent data from being sent out right away.

Reply:
(To clarify, I think it would make a nicer API to allow chaining of linked buffer in calls to `SendByReference` and have each "send done" callback correspond to an `SendByReference` call, but this would require the TCP stack to remember which linked buffer correspond to the same `SendbyReference` call and which don't, which seems quite cumbersome.)

Reply:
@simonlingoogle When you say "disallow appending data" do you mean disallow additional calls to `SendByReference`?

Reply:
> @simonlingoogle When you say "disallow appending data" do you mean disallow additional calls to `SendByReference`?

I meant appending data to the `otLinkedBuffer` that is already sent by `SendByReference`.

Reply:
OK. Can you clarify why this is a bad idea? As we discussed in the meeting it would solve the "repeated one-byte write" problem that @jwhui raised at our meeting today.

Reply:
Sorry, I meant appending data to the `otLinkedBuffer` that is already in transmission (after calling SendByReference, but before ACKed).
If we allow that, "send done" should notify until all data in the `otLinkedBuffer` is sent successfully.
I think this is implementable, but may not be useful. If the previous data is already in transmission, the latter appended data won't be sent in the same segment anyway if the previous data is sent successfully without retransmission, which should be the very likely. 

That's why I think appending data to `otLinkedBuffer` is less meaningful once the `otLinkedBuffer` is in transmission. 

Thoughts?

Reply:
The intention isn't that each `otLinkedBuffer` corresponds to a single TCP segment. The TCP implementation is responsible for breaking up the stream into segments, and the boundaries between TCP segments does not necessarily match the boundary between linked buffers. I think that the logic of how to break up a TCP stream into individual segments belongs in the TCP implementation, not in the application using TCP.

See my comment [here](https://github.com/openthread/openthread/pull/6491#discussion_r616770478) for more details on what a truly zero-copy send path might look like.

Reply:
> (To` clarify, I think it would make a nicer API to allow chaining of linked buffer in calls to SendByReference and have each "send done" callback correspond to an SendByReference call, but this would require the TCP stack to remember which linked buffer correspond to the same SendbyReference call and which don't, which seems quite cumbersome.)

I agree. I think we want to allow back-to-back calls to add more content to send.

I was thinking of same regarding how the "send done" callback work (i.e. we may need to track which `otLinkBuffer` groups where passed together in the implementation which is somehwat messy). How about this, we decouple the two and allow "send-done" callback to include any groups of chained `otLinkBuffer`.

Reply:
That's an option that I considered. However, allowing the "send done" callback to include any groups of chained `otLinkedBuffer` forces the application developer to write code that's resilient to any kind of grouping. It seemed less error prone to have each "send done" callback correspond to a single `otLinkedBuffer`. In this model, if the TCP implementation wants to give 2 linked buffers back to the application, the application would get two separate "send done" callbacks. If we allowed the 2 linked buffers to be linked, then the application would have to handle the case where they are linked AND the case where they are not linked.

Initially, I was thinking along the same lines that you proposed (allowing a single "send done" callback to include a group of chained `otLinkedBuffer`s) but @StuartCheshire convinced me that that's likely to lead to subtle bugs in applications causing them to break in edge cases where the `otLinkedBuffer`s are grouped differently than they expect. I'll let him chime in with more details.

Reply:
I am fine a single `otLinkedBuffer` in "send-done" as well. 

To me the key point is that we document the behaviro in API description. Specially that, user can call `otTcpActiveSendByReference` again after a previous successful call and before "send done" callback. And how/when the "send done" callback will be invoked.

There may other behavior to consider and describe:
- In general, since the current model on tx path is that we expect the `otLinkedBuffers` to be initially created and owned by the app layer (OT API user), we need to think of how to notify it so it can reclaim its buffers in all failure cases.
- For example, what if the socket gets disconnected? Do we invoke "send done" with some failure for each still  remaining `otLinkedBufffer` (to give chance to app layer to reclaim its buffers?) Or a new callback?

Reply:
Thanks. I agree that we should revise the comments describing the functions to clarify the behavior. I'll make a pass where I update the descriptions of the relevant functions to clarify this.

If the socket gets disconnected, we already have the "connection lost" callback. I was thinking that this could be the signal that the application can reclaim all of its `otLinkedBuffer`s. Do you feel that that's sufficient?

Reply:
I think we should allow multiple send calls without having to wait for a "send done" event. I think we should have a callback that indicates how many bytes were acknowledged. This makes it clear that the bytes were acknowledged rather than the socket simply being closed.

Reply:
In the latest update I clarified the comments to explain that you can do multiple send calls without a "send done" event. I also added a callback that tells the application when the peer acknowledges bytes in the send buffer.

Reply:
@abtink Are the updated documentation comments OK?